Implement logging of metrics

joshkraft opened the pull request:
Adds a MetricsLogger class that keeps track of several metrics related to the execution of scripts:

- total successes (ignoring `no-command`'s, which don't run anything)
- scripts that ran
- scripts that were fresh 
- scripts that were restored from cache
- elapsed time

In 'normal' mode, metics are logged on each iteration with at least one 'success'. In 'watch mode', metrics are logged on each _interesting_ iteration with at least one 'success'.

<img width="1132" alt="metrics" src="https://user-images.githubusercontent.com/30248392/220718913-382c5647-ea83-45eb-b309-4fac7c610084.png">


 Addresses issue #525

rictic commented on src/logging/metrics-logger.ts:
```diff
@@ -0,0 +1,112 @@
+      count: 0,
+    },
+    {
+      name: 'Skipped (fresh)',
```
I don't think it'll be very clear from a user's perspective how`Skipped (cached)` is different from `Skipped (fresh)`. Consider instead `Restored from cache`

joshkraft replied:
Agreed - changing to `Restored from cache`

rictic reviewed (commented):
One high level comment: I think printMetrics should be a method on Logger

If we want a MetricsLogger that behaves like DefaultLogger only it prints metrics then we'd give DefaultLogger a printMetrics method that does nothing, and MetricsLogger would be a subclass that overrides it

joshkraft commented:
> One high level comment: I think printMetrics should be a method on Logger
> 
> If we want a MetricsLogger that behaves like DefaultLogger only it prints metrics then we'd give DefaultLogger a printMetrics method that does nothing, and MetricsLogger would be a subclass that overrides it

Hi @rictic - thanks for the feedback. I took your suggestion and refactored this a bit. Would you mind taking a look to make sure what I implemented aligns with that you had in mind? Thanks!

joshkraft commented on src/logging/default-logger.ts:
```diff
@@ -306,4 +306,8 @@ export class DefaultLogger implements Logger {
     }
   }
+
+  printMetrics(): void {
```
When this method is empty, eslint complains:

> Unexpected empty method 'printMetrics'.eslint@typescript-eslint/no-empty-function

This could also be resolved by adding `//eslint-disable-line` or updating the ESLint config, but adding a comment seemed pragmatic.

rictic replied:
+1

joshkraft commented on src/logging/watch-logger.ts:
```diff
@@ -46,6 +48,10 @@ export class WatchLogger implements Logger {
     }
   }
 
+  printMetrics(): void {
```
When this method is empty, eslint complains:

> Unexpected empty method 'printMetrics'.eslint@typescript-eslint/no-empty-function

This could also be resolved by adding `//eslint-disable-line` or updating the ESLint config, but adding a comment seemed pragmatic.

rictic reviewed (approved):
Looks good to me!

rictic commented on src/test/metrics.test.ts:
```diff
@@ -0,0 +1,323 @@
+import {suite} from 'uvu';
```
Please include a copyright header at the top of this file, with a 2023 copyright date

joshkraft replied:
Done!

aomarks replied:
Seems like the copyright header is still missing from this file.

joshkraft replied:
Oops, just pushed up the header to this file.

rictic reviewed (changes requested):
Ok, I've tested it locally, this works great! Just one small thing and it's good to go

rictic commented on src/logging/metrics-logger.ts:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  private _resetMetrics(): void {
+    this._startTime = hrtime();
```
This does the wrong thing in watch mode I think. We end up reporting the time from the end of the last run, rather than the beginning of this run.

So e.g. if you go get coffee for five minutes, then come back and trigger a build that takes three seconds, the elapsed time will be reported as five minutes and three seconds

joshkraft replied:
Nice catch. I just added some code to reset the metrics when a `watch-run-start` event is received, I assume this is when we would want to 'start the timer'?

aomarks replied:
Yes I think this is correct now :+1:

aomarks commented on src/logging/metrics-logger.ts:
```diff
@@ -0,0 +1,128 @@
+   */
+  override log(event: Event): void {
+    // When in watch mode, metrics should reset at the start of each run.
+    if (event.type === 'info' && event.detail === 'watch-run-start') {
```
Not the biggest deal, but it seems a little asymmetrical that this class intrinsically understands when it should reset its metrics, but requires the caller to tell it when to display the metrics.

One idea I had was to make each instance of a `MetricsLogger` only live for the duration of a single iteration. So, in one-shot mode there would be one instance, but in watch mode we would create a new one on each iteration. The benefit of that is that we don't need any reset logic -- we can just use class field initializers to set the counters to 0 and the start time to now.

I'm also fine with it as-is, though.

joshkraft replied:
I think this is a good suggestion, looking into it. I explored this a bit when I got started with this work, but couldn't identify an existing event that signals 'the last service has finished executing', which I assume is when we would want MetricsLogger to automatically print metrics to the console. 

Do you know if there is an existing way to know when all services have finished executing, or would we need a new event (or some other signal)?

aomarks replied:
Yeah, maybe it's not so straightforward as I imagined. This is great as-is, we can always refactor things later if needed.

aomarks reviewed (changes requested):
Awesome work!

thiagowfx commented:
Drive-by: Is it possible to opt-out of metrics? Any flag or switch we can add to do so? (Apologies if this is already documented elsewhere, I could not find a way to do so.)

joshkraft commented:
> Drive-by: Is it possible to opt-out of metrics? Any flag or switch we can add to do so? (Apologies if this is already documented elsewhere, I could not find a way to do so.)

Hi @thiagowfx - this wasn't originally created with any opt-out mechanism. FWIW, the metrics are not aggregated or sent out - it stays local. Is your concern just around reducing the amount of output when executing commands?

thiagowfx commented:
> Is your concern just around reducing the amount of output when executing commands?

Yes, pretty much only that.

A typical `wireit` command in our repository () now looks like this with version `0.10.0`:
<img width="339" alt="image" src="https://github.com/google/wireit/assets/2840106/f611137d-fed8-4ac6-ae96-664e7cb754d4">

...and it's not particularly actionable.

thiagowfx commented:
With `--watch` it's even noisier, e.g.:

<img width="352" alt="image" src="https://github.com/google/wireit/assets/2840106/1e695ba8-a6de-4978-8241-fba245e7a2f2">

rehanvdm commented:
Ticket to disable/opt-out of metrics https://github.com/google/wireit/issues/830
